### PR TITLE
Kronecker Product addition to stdlib_linalg

### PR DESCRIPTION
**Pull request by adenchfi**
Based on this issue, (https://github.com/fortran-lang/stdlib/issues/699)
I decided to go ahead and make a PR with this functionality. In case I go forward with the GSoC program, I figured the sooner the better. 

I basically used the same format as linalg_outer_product for the .fypp source and for the unit tests. I confirmed I can build the stdlib with these additions with cmake, and that all unit tests pass. 

The API and ordering convention used is similar to numpy.kron, though my name is slightly longer. Additionally, because there is both a Kronecker product and a Kronecker sum, I decided to use the full name kronecker_product.

It was suggested to go small with the PRs, and this is my first one, so I have not included kronecker_sum in this PR. 

PS: There is no change to the README, I am not sure why there was anything to commit in that one. I was adding an SSH key to my machine's git and I suppose (?) the commands it had me do forced another commit.

### CODE REVIEW

File: src/stdlib_linalg.fypp [hunk 1]
```diff
@@ -93,6 +94,20 @@ module stdlib_linalg
     #:endfor
   end interface outer_product
 
+  interface kronecker_product
+    !! version: experimental
+    !!
+    !! Computes the Kronecker product of two arrays size M1xN1, M2xN2, returning an (M1*M2)x(N1*N2) array
```
Comment (jvdp1) [inline src/stdlib_linalg.fypp:100]:
```suggestion
    !! Computes the Kronecker product of two arrays of size M1xN1, and of M2xN2, returning an (M1*M2)x(N1*N2) array
```

File: src/stdlib_linalg_kronecker.fypp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      ${t1}$ :: C(size(A,dim=1)*size(B,dim=1),size(A,dim=2)*size(B,dim=2))
+      integer :: m1, n1, m2, n2, maxM1, maxN1, maxM2, maxN2
+      
+      maxM1 = size(A, dim=1)
+      maxN1 = size(A, dim=2)
+      maxM2 = size(B, dim=1)
+      maxN2 = size(B, dim=2)
+      
+      do n1=1, maxN1
+         do m1=1, maxM1
```
Comment (jvdp1) [inline src/stdlib_linalg_kronecker.fypp:21]:
```suggestion
      do n1 = 1, maxN1
         do m1 = 1, maxM1
```

File: test/linalg/test_linalg.fypp [hunk 3]
```diff
@@ -552,6 +562,336 @@ contains
             "trace(h) == sum(c(0:nd:2)) failed.")
 
     end subroutine test_trace_int64
+    
+    subroutine test_kronecker_product_rsp(error)
```
Comment (jvdp1) [inline test/linalg/test_linalg.fypp:566]: I suggest to move all these tests in a fypp loop. This would avoid all these copy-pastes and possible errors/typos.

File: test/linalg/test_linalg.fypp [hunk 4]
```diff
@@ -552,6 +562,336 @@ contains
             "trace(h) == sum(c(0:nd:2)) failed.")
 
     end subroutine test_trace_int64
+    
+    subroutine test_kronecker_product_rsp(error) 
+      !> Error handling
+      type(error_type), allocatable, intent(out) :: error
+      integer, parameter :: m1=1, n1=2, m2=2, n2=3
```
Comment (jvdp1) [inline test/linalg/test_linalg.fypp:569]:
```suggestion
      integer, parameter :: m1 = 1, n1 = 2, m2 = 2, n2 = 3
```

File: test/linalg/test_linalg.fypp [hunk 5]
```diff
@@ -552,6 +562,336 @@ contains
...
+      real(sp), parameter :: tol = 1.e-6
+
+      real(sp) :: A(m1,n1), B(m2,n2)
+      real(sp) :: C(m1*m2,n1*n2), expected(m1*m2,n1*n2), diff(m1*m2,n1*n2)
+
+      integer :: i,j
+
+      do j=1, n1
+         do i=1, m1
```
Comment (jvdp1) [inline test/linalg/test_linalg.fypp:578]:
```suggestion
      do j = 1, n1
         do i = 1, m1
```

File: test/linalg/test_linalg.fypp [hunk 6]
```diff
@@ -552,6 +562,336 @@ contains
...
+            A(i,j) = i*j ! A = [1, 2]
+         end do
+      end do
+
+      do j=1, n2
+         do i=1, m2
```
Comment (jvdp1) [inline test/linalg/test_linalg.fypp:584]:
```suggestion
      do j = 1, n2
         do i = 1, m2
```

File: test/linalg/test_linalg.fypp [hunk 7]
```diff
@@ -552,6 +562,336 @@ contains
...
+            B(i,j) = i*j ! B = [[1, 2, 3], [2, 4, 6]]
+         end do
+      end do
+
+      C = kronecker_product(A,B)
+      expected = transpose(reshape([1,2,3, 2,4,6, 2,4,6, 4, 8, 12], [m2*n2, m1*n1]))
```
Comment (jvdp1) [inline test/linalg/test_linalg.fypp:590]: This matrix `expected` could be a `real, parameter`

File: src/stdlib_linalg_kronecker.fypp [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+            ! We use the numpy.kron convention for ordering of the matrix elements
```
Comment (jvdp1) [inline src/stdlib_linalg_kronecker.fypp:22]: I suggest that you refer to a reference that does not depend on a language. For example, you could refer to Wikipedia (that follows the same ordering).

**Review by jvdp1 — CHANGES_REQUESTED** [review]
Thank you for this PR. I only have minor comments regarding the code.
Also, could you edit the specs and add this implementation, please?

File: test/linalg/test_linalg.fypp (hunk 3 above)
Comment (adenchfi) [reply test/linalg/test_linalg.fypp:566]: I see. I could not find the fypp loop for the other tests to be generated. The reason I assumed the other tests weren't in a loop is because there seemed to be special behavior for checking if there is quad_precision. I did, however, generate the subroutines using an fypp loop before adding the if-statements for checking for quad precision. 
Do you have a suggestion for me for this? If I loop over all supported types normally, it won't generate quad_precision when someone makes stdlib?

File: src/stdlib_linalg.fypp (hunk 1 above)
Comment (adenchfi) [reply src/stdlib_linalg.fypp:100]: "Also, could you edit the specs and add this implementation, please?"
I am not sure what you mean. Do you mean to commit the suggestion you made? What implementation do you mean I should add? I have an stdlib_linalg_kronecker.fypp file with the implementation in there?

File: src/stdlib_linalg_kronecker.fypp (hunk 8 above)
Comment (adenchfi) [reply src/stdlib_linalg_kronecker.fypp:22]: Sure, I'll do that.

File: test/linalg/test_linalg.fypp (hunk 7 above)
Comment (adenchfi) [reply test/linalg/test_linalg.fypp:590]: I could make it a parameter. Should I break with the outer_product convention to do so?

File: src/stdlib_linalg.fypp (hunk 1 above)
Comment (adenchfi) [reply src/stdlib_linalg.fypp:100]: Ah shit, I can't read the contributor workflow document apparently.

How should I create the spec? I don't see that in tabulated in the workflow. Should I just make a LaTeX/PDF document?

Comment (adenchfi) [reply src/stdlib_linalg.fypp:100]: Ah, figured it out. Yeah, I'll work on a spec.

File: test/linalg/test_linalg.fypp (hunk 7 above)
Comment (jvdp1) [reply test/linalg/test_linalg.fypp:590]: I don't think there is a convention for such things.

File: test/linalg/test_linalg.fypp (hunk 3 above)
Comment (jvdp1) [reply test/linalg/test_linalg.fypp:566]: Just generate a loop with the REAL_KINDS_TYPES. If QP is not supported, it it not added to the latter variable. See [common.fypp](https://github.com/fortran-lang/stdlib/blob/master/src/common.fypp) for the details and also [test_mean.fypp](https://github.com/fortran-lang/stdlib/blob/master/test/stats/test_mean.fypp) for an example.

File: src/stdlib_linalg.fypp (hunk 1 above)
Comment (adenchfi) [reply src/stdlib_linalg.fypp:100]: I pushed a spec and associated example.

File: doc/specs/stdlib_linalg.md [hunk 9]
```diff
@@ -160,6 +160,37 @@ Returns a rank-2 array equal to `u v^T` (where `u, v` are considered column vect
 {!example/linalg/example_outer_product.f90!}
 ```
 
+## `kronecker_product` - Computes the kronecker product of two rank-2 arrays
```
Comment (jvdp1) [inline doc/specs/stdlib_linalg.md:163]:
```suggestion
## `kronecker_product` - Computes the Kronecker product of two rank-2 arrays
```

File: example/linalg/example_kronecker_product.f90 [hunk 10]
```diff
@@ -0,0 +1,25 @@
+program example_kronecker_product
+  use stdlib_linalg, only: kronecker_product
+  implicit none
+  integer, parameter :: m1=1, n1=2, m2=2, n2=3
```
Comment (jvdp1) [inline example/linalg/example_kronecker_product.f90:4]:
```suggestion
  integer, parameter :: m1 = 1, n1 = 2, m2 = 2, n2 = 3
```

File: example/linalg/example_kronecker_product.f90 [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  real :: A(m1, n1), B(m2,n2)
+  real, allocatable :: C(:,:)
+
+  do j=1, n1
+     do i=1, m1
```
Comment (jvdp1) [inline example/linalg/example_kronecker_product.f90:9]:
```suggestion
  do j = 1, n1
     do i = 1, m1
```

File: example/linalg/example_kronecker_product.f90 [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        A(i,j) = i*j ! A = [1, 2]
+     end do
+  end do
+  
+  do j=1, n2
+     do i=1, m2      ! B = [ 1, 2, 3 ]
```
Comment (jvdp1) [inline example/linalg/example_kronecker_product.f90:15]:
```suggestion
  do j = 1, n2
     do i = 1, m2      ! B = [ 1, 2, 3 ]
```

**Review by jvdp1 — CHANGES_REQUESTED** [review]
Specs look good. Thank you.

File: test/linalg/test_linalg.fypp (hunk 3 above)
Comment (adenchfi) [reply test/linalg/test_linalg.fypp]: Got it. I condensed the test into a single fypp loop and also made the expected array a parameter. Rebuilt and reran the tests, all passed. Checked the tests .f90 file, all tests seem to have generated fine.

File: test/linalg/test_linalg.fypp (hunk 7 above)
Comment (adenchfi) [reply test/linalg/test_linalg.fypp]: In my push I made `expected` a parameter.

File: example/linalg/example_kronecker_product.f90 [hunk 13]
```diff
@@ -0,0 +1,25 @@
+program example_kronecker_product
+  use stdlib_linalg, only: kronecker_product
+  implicit none
+  integer, parameter :: m1 = 1, n1 = 2, m2 = 2, n2 = 3
```
Comment (14NGiestas) [inline example/linalg/example_kronecker_product.f90:4]:
```suggestion
  integer, parameter :: m1 = 1, n1 = 2, m2 = 2, n2 = 3
  integer :: i, j
```

File: doc/specs/stdlib_linalg.md [hunk 14]
```diff
@@ -160,6 +160,37 @@ Returns a rank-2 array equal to `u v^T` (where `u, v` are considered column vect
...
+
+### Syntax
+
+`C = [[stdlib_linalg(module):kronecker_product(interface)]](A, B)`
+
+### Arguments
+
+`A`: Shall be a rank-2 array with dimensions M1, N1
+
+`v`: Shall be a rank-2 array with dimensions M2, N2
```
Comment (jvdp1) [inline doc/specs/stdlib_linalg.md:181]:
```suggestion
`B`: Shall be a rank-2 array with dimensions M2, N2
```

File: test/linalg/test_linalg.fypp [hunk 15]
```diff
@@ -48,6 +48,16 @@ contains
...
+            new_unittest("kronecker_product_rsp", test_kronecker_product_rsp), &
+            new_unittest("kronecker_product_rdp", test_kronecker_product_rdp), &
+            new_unittest("kronecker_product_rqp", test_kronecker_product_rqp), &
+            new_unittest("kronecker_product_csp", test_kronecker_product_csp), &
+            new_unittest("kronecker_product_cdp", test_kronecker_product_cdp), &
+            new_unittest("kronecker_product_cqp", test_kronecker_product_cqp), &
+            new_unittest("kronecker_product_int8", test_kronecker_product_iint8), &
+            new_unittest("kronecker_product_int16", test_kronecker_product_iint16), &
+            new_unittest("kronecker_product_int32", test_kronecker_product_iint32), &
+            new_unittest("kronecker_product_int64", test_kronecker_product_iint64), &
```
Comment (jvdp1) [inline test/linalg/test_linalg.fypp:60]: All of these should be generated by  `fypp` too. Otherwise there might be issues with `qp` precision (i.e. if it is supported or not)

File: example/linalg/example_kronecker_product.f90 (hunk 13 above)
Comment (adenchfi) [reply example/linalg/example_kronecker_product.f90:4]: Ah, thank you.

File: test/linalg/test_linalg.fypp (hunk 15 above)
Comment (adenchfi) [reply test/linalg/test_linalg.fypp]: Alright. I made the change, and the build built and all tests passed. I checked the built `build/test/test_linalg.f90` and it seemed to have them listed. I suppose we could really clean up the linalg test source if all the other linalg tests were fypp generated like this.

**Comment by adenchfi** [comment]
I believe there's nothing else for this particular PR?

**Review by jvdp1 — APPROVED** [review]
LGTM.  I have no additional comments.
hank you for this PR.

**Comment by adenchfi** [comment]
> LGTM. I have no additional comments. Thank you for this PR.

Thanks for guiding me through the process! I will work on the Kronecker sum as well sometime.

**Comment by jvdp1** [comment]
Thank you @adenchfi 
@14NGiestas Can you have a final look, please? Then I will merge it!

**Comment by 14NGiestas** [comment]
@jvdp1 LGTM, whould"the code doesn't work for matrices arbitrarily indexed"

**Comment by 14NGiestas** [comment]
@jvdp1 LGTM - I'm just wondering if it works with arbitrarily indexed matrices (as an edge case for tests).

File: src/stdlib_linalg_kronecker.fypp [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+  implicit none
+
+contains
+
+  #:for k1, t1 in RCI_KINDS_TYPES
+    pure module function kronecker_product_${t1[0]}$${k1}$(A, B) result(C)
+      ${t1}$, intent(in) :: A(:,:), B(:,:)
+      ${t1}$ :: C(size(A,dim=1)*size(B,dim=1),size(A,dim=2)*size(B,dim=2))
+      integer :: m1, n1, m2, n2, maxM1, maxN1, maxM2, maxN2
```
Comment (14NGiestas) [inline src/stdlib_linalg_kronecker.fypp:13]:
```suggestion
      integer :: m1, n1, maxM1, maxN1, maxM2, maxN2
```
Almost forgot this one.
Unused variables here, maybe just use just `m`, `n`

**Comment by 14NGiestas** [comment]
> > @jvdp1 LGTM - I'm just wondering if it works with arbitrarily indexed matrices (as an edge case for tests).
> 
> I am not sure what you mean. Somehting like:
> 
> ```fortran
> real :: mat(-5:7, 9:11)
> ```

That was what I had in mind (just to be picky), but I tested and it works anyway:

```fortran
program main
    real :: A(0:2) = [0, 1, 2]
    call m(A)
contains
    subroutine m(A)
        real :: A(:)
        print*, lbound(A), ubound(A) ! 1 3
    end subroutine
end program
```

**Comment by jvdp1** [comment]
> > > @jvdp1 LGTM - I'm just wondering if it works with arbitrarily indexed matrices (as an edge case for tests).
> > 
> > 
> > I am not sure what you mean. Somehting like:
> > ```fortran
> > real :: mat(-5:7, 9:11)
> > ```
> 
> That was what I had in mind (just to be picky), but I tested and it works anyway:
> 
> ```fortran
> program main
>     real :: A(0:2) = [0, 1, 2] ! or A(-1:1) and so on
>     call m(A)
> contains
>     subroutine m(A)
>         real :: A(:)
>         print*, lbound(A), ubound(A) ! 1 3
>     end subroutine
> end program
> ```

Thank you. I'll commit your last change and merge it.
Tahnk @adenchfi for your contributions
